### PR TITLE
Lazy-init formatter resolvers in MessagePackHubProtocolOptions

### LINKED ISSUE #10309: Lazy-initialize MessagePackHubProtocolOptions
We should lazy-init the formatter resolvers in MessagePackHubProtocolOptions. The default set of resolvers trigger a static constructor that throws on AOT environments (IL2CPP for Unity for example). If we lazy init it, an AOT-based app can call the setter to replace the list of formatters with an AOT-friendly set without risking an exception due to trying to initialize the non-AOT-friendly ones.

### PR DESCRIPTION
**Pull request by mikaelm12**
Fixes: https://github.com/aspnet/AspNetCore/issues/10309

### CODE REVIEW

File: src/SignalR/common/Protocols.MessagePack/src/MessagePackHubProtocolOptions.cs [hunk 1]
```diff
@@ -9,6 +9,23 @@ namespace Microsoft.AspNetCore.SignalR
...
-        public IList<IFormatterResolver> FormatterResolvers { get; set; } = MessagePackHubProtocol.CreateDefaultFormatterResolvers();
+        private IList<IFormatterResolver> _formatterResolvers;
+
+        public IList<IFormatterResolver> FormatterResolvers
+        {
+            get
+            {
+                if (_formatterResolvers == null)
+                {
+                    _formatterResolvers = MessagePackHubProtocol.CreateDefaultFormatterResolvers();
```
Comment (BrennanConroy) [inline src/SignalR/common/Protocols.MessagePack/src/MessagePackHubProtocolOptions.cs]: Lets add a comment explaining why we're doing this

**Review by BrennanConroy — APPROVED** [review]

File: src/SignalR/common/Protocols.MessagePack/src/MessagePackHubProtocolOptions.cs [hunk 2]
```diff
@@ -9,6 +9,25 @@ namespace Microsoft.AspNetCore.SignalR
...
+                    // The default set of resolvers trigger a static constructor that throws on AOT environments.
+                    // This gives users the chance to use an AOT friendly formatter.
+                    _formatterResolvers = MessagePackHubProtocol.CreateDefaultFormatterResolvers();
+                }
+
+                return _formatterResolvers;
+            }
+            set
+            {
+                _formatterResolvers = value;
```
Comment (halter73) [inline src/SignalR/common/Protocols.MessagePack/src/MessagePackHubProtocolOptions.cs]: I'm assuming there's no reason that anyone would actually want to set this to null, right?

**Review by halter73 — APPROVED** [review]

**Comment by mikaelm12** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=213777&view=logs

File: src/SignalR/common/Protocols.MessagePack/src/MessagePackHubProtocolOptions.cs (hunk 2 above)
Comment (mikaelm12) [reply src/SignalR/common/Protocols.MessagePack/src/MessagePackHubProtocolOptions.cs]: > I'm assuming there's no reason that anyone would actually want to set this to null, right?

Right. I don't think that's a use case we'll see.

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/2614
https://github.com/aspnet/AspNetCore-Internal/issues/2629

**Pull request merged**